akka-contrib/src/main/scala/akka/contrib/pattern/TransparentExponentialBackoffSupervisor.scala

```diff
@@ -0,0 +1,206 @@
+ * `sender()` `ActorRef` from the child response may eventually not be able to communicate with
+ * the stored `ActorRef`. In general all messages to the child should be directed through the
+ * [[TransparentExponentialBackoffSupervisor]].
+ *
+ * An example of where this supervisor might be used is when you may have an actor that is
+ * responsible for continuously polling on a server for some resource that sometimes may be down.
+ * Instead of hammering the server continuously when the resource is unavailable, the actor will
+ * be restarted with an exponentially increasing back off until the resource is available again.
+ */
+class TransparentExponentialBackoffSupervisor(
+  props: Props,
+  decider: Option[Decider],
+  minBackoff: FiniteDuration,
+  maxBackoff: FiniteDuration,
+  randomFactor: Double)
+  extends Actor
+  with Stash
+  with ActorLogging {
+
+  import TransparentExponentialBackoffSupervisor._
+  import context._
+
+  override val supervisorStrategy = OneForOneStrategy(maxNrOfRetries = -1) {
+    case ex ⇒
+      val defaultDirective: Directive =
+        super.supervisorStrategy.decider.applyOrElse(ex, (_: Any) ⇒ Escalate)
+      val maybeDirective: Option[Directive] = decider
+        .map(_.applyOrElse(ex, (_: Any) ⇒ defaultDirective))
+
+      // Get the directive from the specified decider or fallback to
+      // the default decider.
+      // Whatever the final Directive is, we will translate all Restarts
+      // to our own Restarts, which involves stopping the child.
+      maybeDirective.getOrElse(defaultDirective) match {
+        case Restart ⇒
+          self ! RestartChild
+          Stop
+        case other ⇒ other
+      }
+  }
+
+  // Initialize by starting up and watching the child
+  self ! RestartChild
+
+  def receive = {
+    case RestartChild ⇒
+      val childRef = actorOf(props)
+      watch(childRef)
+      unstashAll()
+      become(watching(childRef, 0))
+    case _ ⇒ stash()
+  }
+
+  // Steady state
+  def watching(childRef: ActorRef, numRestarts: Int): Receive = {
+    case RestartChild ⇒
+      log.debug("Waiting on child termination to restart")
+      become(waitingToRestart(childRef, numRestarts))
+    case ResetRestartCount(last) ⇒
+      if (last == numRestarts) {
+        log.debug(s"Last restart count [$last] matches current count; resetting")
+        become(watching(childRef, 0))
+      } else {
+        log.debug(s"Last restart count [$last] does not match the current count [$numRestarts]")
+      }
+    case Terminated(`childRef`) ⇒
+      log.debug(s"Terminating, because child [$childRef] terminated itself")
+      stop(self)
+    case Terminated(other) ⇒
+      log.debug(s"Ignoring unexpected Terminated message from: $other")
+    case msg ⇒
+      childRef.forward(msg)
+  }
+
+  def calculateDelay(numRestarts: Int): FiniteDuration = {
+    // Source:
+    // https://github.com/akka/akka/blob/v2.4.0/akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala
```

Comment:
Sure, I can refactor the existing impl in the `BackoffSupervisor` into a function in the companion object and make my supervisor make use of that function.